### DISCUSSION TITLE
How to get or set other micro service's insecure secret from application service sdk?

### DISCUSSION THREAD

**Discussion by jiekechoo** [edgexfoundry/community#258] [Q&A] [answered]
The onvif-camera micro service needs to set a secret for the camera onvif protocol access, with my custom app-service, I need to get and set the onvif service's secret, because the custom app-service is the only way to communicate with the cloud.

I think that there are two ways for this:
1. write a `OnvifClient` like the `DeviceClient` in the `app-function-sdk-go`, using the `onvif-camera` service's API ;
2. Is there any api or function that I can set the secret for `onvif-camera` service  in directly?

**Comment by lenny-goodell** [comment]
Best way is to use the Config Provider interface. 

The scripts here are do this from the command-line assuming Consul is the Config Provider.
https://github.com/edgexfoundry/device-onvif-camera/tree/main/bin

**Reply by jiekechoo** [reply]
@lenny-intel , thanks a lot.

I don't want to make a big application, just a simple one.

Maybe calling a http rest of the onvif service , like the `DeviceClient().Add()`.

**Reply by lenny-goodell** [reply]
Actual you can now use the device service's /secret API for both secure and non-secure secrets.

See /secret API swagger here: https://docs.edgexfoundry.org/3.1/microservices/device/ApiReference/

**Reply by jiekechoo** [reply]
Yes. But how to make a request in the `app-application-sdk-go` by build-in function?

**Reply by lenny-goodell** [reply]
This call would never go into the App SDK. 

If you have a need for this, you would put it in your custom App Service and simply make a REST call to the appropriate device service URL.

**Reply by jiekechoo** [reply]
Yes, I use the function `utils.PostRequestWithRawData` for making a REST call to the `onvif-camera` device service. https://github.com/edgexfoundry/go-mod-core-contracts/blob/b0286c308de37bd3d0f6157ee2560331f20d5af8/clients/http/utils/request.go#L113-L133

and, I added  the `device-onvif-camera` connection info to `configuration.yaml`, how to get the device service's `baseUrl` in the application serivce? Can you give me some reference? 
```yaml
...
  device-onvif-camera:
    Protocol: 'http'
    Host: 'localhost'
    Port: 59984
    UserMessageBus: true
...
```

it's in the application service `config`:
```
// http://localhost:62801/api/v3/config

{
  "apiVersion": "v3",
...
    "Clients": {
      "core-command": {
        "Host": "localhost",
        "Port": 59882,
        "Protocol": "http",
        "UseMessageBus": true
      },
...
      "device-onvif-camera": {
        "Host": "localhost",
        "Port": 59984,
        "Protocol": "http",
        "UseMessageBus": false
      }
    }
...
```

**Reply by jiekechoo** [reply, marked as answer]
I got it.
The RegistryClient should be enabled before call the `GetServiceEndpoint`.
```
endPoint, err := registryClient.GetServiceEndpoint("device-onvif-camera")
```
